critwall/src/main/java/org/opencadc/critwall/FileSync.java

```diff
@@ -171,16 +196,40 @@ public void run() {
 
                     FileSyncJob fsj = new FileSyncJob(curArtifact.getURI(), this.locatorService,
                         this.storageAdapter, this.jobArtifactDAO);
-                    fsj.run();
+
+                    log.debug("creating file sync job: " + curArtifact.getURI());
+
+                    threadPool.addTask(fsj);
+                    log.debug("added FileSyncJob to thread pool.");
+
                 }
             }
+
         } catch (Exception e) {
-            // capture database errors here and quit if there's a failure
+            log.info("Thread pool error", e);
```

Comment:
OK. error handling still TBD.

Once you finish the `while (bucketSelector.hasNext())` loop, the main thread can exit. As a well commented *temporary* hack, you can add something like
```
// HACK: temporarily keep running until all jobs are completed
while (!jobQueue.isEmpty()) {
   Thread.sleep(6000L);
}
```

Reply:
!jobQueue.isEmpty() isn't going to be a sufficient check - there are threads still busy putting the files to local storage when the main thread stops.

Keeping the Thread.sleep however. Should the value be a parameter?